src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDataTable.java

```diff
@@ -207,92 +220,86 @@ private void internalAddrow(JDBCType jdbcType, Object val, Object[] rowValues,
-                            null);
-                }
-                JavaType javaType = JavaType.of(val);
-                internalJDBCType = javaType.getJDBCType(SSType.UNKNOWN, jdbcType);
-                internalAddrow(internalJDBCType, val, rowValues, pair);
-                break;
-            default:
-                MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_unsupportedDataTypeTVP"));
-                Object[] msgArgs = {jdbcType};
-                throw new SQLServerException(null, form.format(msgArgs), null, 0, false);
+                    rowValues[pair.getKey()] = bd;
+                    break;
+
+                case DOUBLE:
+                    rowValues[pair.getKey()] = (val instanceof Double) ? val : Double.parseDouble(val.toString());
+                    break;
+
+                case FLOAT:
+                case REAL:
+                    rowValues[pair.getKey()] = (val instanceof Float) ? val : Float.parseFloat(val.toString());
+                    break;
+
+                case TIMESTAMP_WITH_TIMEZONE:
+                case TIME_WITH_TIMEZONE:
+                case DATE:
+                case TIME:
+                case TIMESTAMP:
+                case DATETIMEOFFSET:
+                case DATETIME:
+                case SMALLDATETIME:
+                    // Sending temporal types as string. Error from database is thrown if parsing fails
+                    // no need to send precision for temporal types, string literal will never exceed
+                    // DataTypes.SHORT_VARTYPE_MAX_BYTES
+
+                    // java.sql.Date, java.sql.Time and java.sql.Timestamp are subclass of java.util.Date
+                    if (val instanceof java.util.Date || val instanceof microsoft.sql.DateTimeOffset
+                            || val instanceof OffsetDateTime || val instanceof OffsetTime)
+                        rowValues[pair.getKey()] = val.toString();
+                    else
+                        rowValues[pair.getKey()] = val;
+                    break;
+
+                case BINARY:
+                case VARBINARY:
+                case LONGVARBINARY:
+                    nValueLen = ((byte[]) val).length;
+
+                    if (nValueLen > currentColumnMetadata.precision) {
+                        currentColumnMetadata.precision = nValueLen;
+                        columnMetadata.put(pair.getKey(), currentColumnMetadata);
+                    }
+                    rowValues[pair.getKey()] = val;
+                    break;
+
+                case CHAR:
+                case VARCHAR:
+                case NCHAR:
+                case NVARCHAR:
+                case LONGVARCHAR:
+                case LONGNVARCHAR:
+                case SQLXML:
+                    if (val instanceof UUID)
+                        val = val.toString();
+                    nValueLen = (2 * ((String) val).length());
```

Comment:
Why are we multiplying length by 2 for `CHAR` type? I know you didn't introduce this, but seems like an issue to me.

Reply:
I wouldn't modify unless this is reported as issue. This could be by design.

Reply:
Probably because CHAR and VARCHAR bytes per character is half the size of the others. The length value here is meant to represent the number of characters and not the length of the buffer. I had to do something similar for clob vs nclob.

Reply:
not really, `CHAR`, `VARCHAR` are single-byte types, meaning their length and number of characters are the same.

Reply:
We'll keep it in backlog, could be investigated later.